### ISSUE TITLE
Performance regression with arrays

### ISSUE THREAD

**Issue by dvarrazzo** [psycopg/psycopg#359]
Not sure if it's related, but I'm seeing significantly reduced performance when using psycopg3 as compared to psycopg2, in the absence of Pandas. On this end most queries in the app I tried to migrate return on average 2x-3x slower than before.

Below is a simple repro that runs ~5x more slowly if I change from psycopg2 to psycopg. I tried with both psycopg binary and c, version 3.0.16. psycopg2 binary, version 2.8.6.

```
def f():
    con = psycopg2.connect([CONNECTION_STRING])
    cur = con.cursor()
    cur.execute("select array[0,1,2,3,4,5,6,7,8,9] from generate_series(0, 100000) as g(r)")
    return(list(iter(cur)))
```

_Originally posted by @amachanic in https://github.com/psycopg/psycopg/issues/316#issuecomment-1228939941_

**Comment by dvarrazzo** [comment]
That's confirmed. The slower performance comes from the fact that parsing arrays is implemented in Python in psycopg 3, whereas in psycopg2 is in C.

Now that we have tweaked the Python implementation and we are comfortable that it is correct, it should be easy to use it as a reference to implement a C version.

Thank you for your report!

**Comment by WoLpH** [comment]
Not sure if it's something you're interested in but it's related :)

I've written a much faster array loader for psycopg 3 that can immediately load the data into `numpy`: https://github.com/WoLpH/speedups

**Comment by dvarrazzo** [comment]
Thank you, @WoLpH. We are actually interested in providing dumpers/loaders for numpy arrays (#336) and someone is already working on numpy basic types (#192). 

In order to consider this issue closed we can convert the Python loaders (and dumpers, if it's worth) to Cython, which will be a massive speedup because the algorithm works mostly on C structures which don't require the Python unpack overhead. We prefer to not introduce a numpy dependency for this operation.

**Comment by dvarrazzo** [comment]
Releasing a fix in psycopg 3.1.5.
